### PR TITLE
KAFKA-13527: Add top-level error code field to DescribeLogDirsResponse

### PR DESCRIPTION
**Pull request by mimaison**
Implements KIP-784

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by tombentley — APPROVED** [review]

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 1]
```diff
@@ -2681,8 +2681,11 @@ public void handleResponse(AbstractResponse abstractResponse) {
                     if (descriptions.size() > 0) {
                         future.complete(descriptions);
                     } else {
-                        // descriptions will be empty if and only if the user is not authorized to describe cluster resource.
-                        future.completeExceptionally(Errors.CLUSTER_AUTHORIZATION_FAILED.exception());
+                        // Up to v3 DescribeLogDirsResponse did not have an error code field, hence it defaults to None
+                        Errors error = response.data().errorCode() == Errors.NONE.code()
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:2685]: Should we extend unit tests in `KafkaAdminTest` to cover this change?

File: clients/src/main/resources/common/message/DescribeLogDirsRequest.json [hunk 2]
```diff
@@ -19,7 +19,7 @@
   "listeners": ["zkBroker", "broker"],
   "name": "DescribeLogDirsRequest",
   // Version 1 is the same as version 0.
-  "validVersions": "0-2",
+  "validVersions": "0-3",
```
Comment (dajac) [inline clients/src/main/resources/common/message/DescribeLogDirsRequest.json:22]: nit: Could we also put a comment for version 3?

Comment (mimaison) [reply clients/src/main/resources/common/message/DescribeLogDirsRequest.json:22]: Yes, good idea. Done!

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java (hunk 1 above)
Comment (mimaison) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:2685]: Absolutely, I've updated the tests.

**Comment by mimaison** [comment]
@dajac Do you want to take another look? Thanks

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java (hunk 1 above)
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:2685]: @mimaison What would happen if DescribeLogDirsResponse v3 actually returns NONE as an error (the default value)? It seems to be that we need to take the version into account here as well.

File: clients/src/main/java/org/apache/kafka/common/requests/DescribeLogDirsResponse.java [hunk 3]
```diff
@@ -52,6 +52,7 @@ public int throttleTimeMs() {
     @Override
     public Map<Errors, Integer> errorCounts() {
         Map<Errors, Integer> errorCounts = new HashMap<>();
+        errorCounts.put(Errors.forCode(data.errorCode()), 1);
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/common/requests/DescribeLogDirsResponse.java:55]: Do we need to cover it in `testErrorCountsIncludesNone`?

**Review by dajac — COMMENTED** [review]
@mimaison Thanks for the updates. Don't we need to update the broker side as well?

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java (hunk 1 above)
Comment (mimaison) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:2685]: With v3, if `NONE` is returned there should be some logdirs in the response so we should have entered the if condition above.

File: clients/src/main/java/org/apache/kafka/common/requests/DescribeLogDirsResponse.java (hunk 3 above)
Comment (mimaison) [reply clients/src/main/java/org/apache/kafka/common/requests/DescribeLogDirsResponse.java:55]: Yes, good point. I've pushed an update to cover it.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java (hunk 1 above)
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:2685]: That makes sense, thanks.

**Comment by dajac** [comment]
@mimaison Should we update `KafkaApis` to correct fill the error code when v3 is used by the client? At the moment, the field is not used at all.

**Comment by mimaison** [comment]
> @mimaison Should we update `KafkaApis` to correct fill the error code when v3 is used by the client? At the moment, the field is not used at all.

@dajac Yes absolutely! I'm not sure why I did not include it earlier. I've just pushed an update

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 4]
```diff
@@ -223,7 +223,7 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
       .find(x => x.topicName == tp.topic).get.partitions.asScala
       .find(p => p.partitionIndex == tp.partition).get.errorCode)),
     ApiKeys.DESCRIBE_LOG_DIRS -> ((resp: DescribeLogDirsResponse) =>
-      if (resp.data.results.size() > 0) Errors.forCode(resp.data.results.get(0).errorCode) else Errors.CLUSTER_AUTHORIZATION_FAILED),
+      Errors.forCode(if (resp.data.results.size() > 0) resp.data.results.get(0).errorCode() else resp.data.errorCode())),
```
Comment (dajac) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: nit: I would remove the parenthesis after `errorCode` to stay consistent with the surrounding code style.

**Review by dajac — APPROVED** [review]
LGTM, thanks. I left a small nit. Feel free to merge once addressed.

**Pull request merged**